### DISCUSSION TITLE
[Question] How to get beautiful and auto-formated ranges?

### DISCUSSION THREAD

**Discussion by kunaltyagi** [ericniebler/range-v3#1634] [General]
My current clang-format configuration produces some ugly looking code, mainly because it splits the logic written using ranges in intuitive places.

Eg:
```cpp
    return some_map | rv::filter([&](const auto& kv) {
        return kv.first == key; }) | rv::values | rv::filter([&](const auto &x) {
        return x %2; });
```

Does anyone have good tips for good-looking code? Any options in clang-format I should set?

Using NOLINT (via clang-format off/on) is tedious and using comments works, but it breaks the flow and defeats the purpose.

A sample clang-format that splits at `|` (doesn't matter if `|` is end of line or start of new line) would be awesome.

**Comment by christianbrugger** [comment]
Two trick I know of:

1. add this to `.clang-format` config. This makes, that the pipes appear in the next line always:
```yaml
BreakBeforeBinaryOperators: All
```

2. Add emtpy  linecomments to force breaking for short segments, if desired.
```c++
auto test() {
  return some_map //
         | rv::filter([&](const auto &kv) { return kv.first == key; })
         | rv::values //
         | rv::filter([&](const auto &x) { return x % 2; });
}
```

You can try this out here:
https://zed0.co.uk/clang-format-configurator/

**Reply by kunaltyagi** [reply]
I've been using the 2nd trick but it does get tedious after a while, specially with small range elements like values and projections. It does work fine for filters :)

**Reply by kunaltyagi** [reply]
I might use both with the break applied only for NonAssignment instead of All. Thanks
